File: 1415. The k-th Lexicographical String of All Happy Strings of Length n.java
class Solution {
    
    String ans;
    int k;
    int n;
    public String getHappyString(int n, int k) {
        this.ans = "";
        this.k = k;
        this.n = n;

        dfs(new char[n], 0);        
        return ans;       
    }

    public boolean dfs(char[] curr, int idx){
        if(idx == n){
            if(k == 1){
                ans = new String(curr);
                return true;
            }
            k--;
            return false;
        }

        for(int i=0; i<3; i++){
            char c = (char)(i + 'a');
            if(idx == 0 || curr[idx-1] != c){
                char pc = curr[idx];
                curr[idx] = c;
                if(dfs(curr, idx+1)){
                    return true;
                }
                curr[idx] = pc;
            }            
        }

        return false;
    }
}

File: 1415. The k-th Lexicographical String of All Happy Strings of Length n.rs
impl Solution {
    pub fn get_happy_string(n: i32, k: i32) -> String {
        let mut count = 0;
        let mut current_s = String::with_capacity(n as usize);

        fn backtrack(
            n: usize, 
            k: i32, 
            count: &mut i32, 
            current_s: &mut String
        ) -> Option<String> {
            if current_s.len() == n {
                *count += 1;
                if *count == k {
                    return Some(current_s.clone());
                }
                return None;
            }

            for c in ['a', 'b', 'c'] {
                if current_s.is_empty() || current_s.chars().last().unwrap() != c {
                    current_s.push(c);
                    
                    if let Some(result) = backtrack(n, k, count, current_s) {
                        return Some(result);
                    }
                    
                    current_s.pop();
                }
            }
            None
        }

        backtrack(n as usize, k, &mut count, &mut current_s).unwrap_or_default()
    }
}
